utils/vpr_pbtype_arch_wrapper.py

```diff
@@ -0,0 +1,404 @@
+    >>> print(grid_format(t))
+       01234
+     0 .....
+     1 .I...
+     2 .....
+     3 .....
+
+    >>> t = grid_new(5, 4)
+    >>> grid_place_in_column(t, 1, ['I', 'I'])
+    >>> grid_place_in_column(t, 3, ['O'])
+    >>> print(grid_format(t))
+       01234
+     0 .....
+     1 .I.O.
+     2 .I...
+     3 .....
+
+    """
+    width, height = grid_size(tiles)
+
+    start = math.floor((height - len(values)) / 2)
+    for i in range(0, len(values)):
+        tiles[(x, start + i)] = values[i]
+
+
+def grid_generate(input_pins: List[str], output_pins: List[str]) -> GridDict:
+    """Generate a grid dict to fit a set of input_pins and output_pins.
+
+    Generates a 5 width grid with following columns;
+     Column 0 - One input tile per input pins.
+     Column 1 - padding
+     Column 2 - A single tile.
+     Column 3 - padding
+     Column 4 - One output tile per output pins.
+
+    """
+    height = max(len(input_pins), len(output_pins)) + 2
+    width = len(['I', '.', 'T', '.', 'O'])
+    tiles = grid_new(width, height)
+    grid_place_in_column(tiles, 0, ['I'] * len(input_pins))
+    grid_place_in_column(tiles, 2, ['T'] * (height - 2))
+    grid_place_in_column(tiles, 4, ['O'] * len(output_pins))
+    return tiles
+
+
+def layout_xml(arch_xml: ET.Element, pbtype_xml: ET.Element):
+    """Generate a `<layout>` with IBUF and OBUF to match given pb_type."""
+
+    pbtype_name, clocks, inputs, outputs = ports(pbtype_xml)
+
+    finputs = [s for s, d in flatten(clocks + inputs)]
+    foutputs = [s for s, d in flatten(outputs)]
+
+    tiles = grid_generate(finputs, foutputs)
+    width, height = grid_size(tiles)
+
+    layouts = arch_xml.find("layout")
+    l = ET.SubElement(
+        layouts,
+        "fixed_layout",
+        {
+            "name": "device",
+            # FIXME: See https://github.com/verilog-to-routing/vtr-verilog-to-routing/issues/277
+            #"width":  str(width),
+            #"height":  str(height),
+            "width": str(max(width, height)),
+            "height": str(max(width, height)),
+        },
+    )
+    l.append(ET.Comment('\n' + grid_format(tiles) + '\n'))
+
+    for x, y in tiles.keys():
+        v = tiles[(x, y)]
+        if v == '.':
+            continue
+        elif v == 'I':
+            t = 'IBUF'
+        elif v == 'O':
+            t = 'OBUF'
+        elif v == 'T':
+            # FIXME: Is this needed?
+            if y > 1:
+                continue
+            t = 'TILE'
+        else:
+            raise Exception("Unknown tile type {}".format(v))
+        ET.SubElement(
+            l, "single", {
+                "type": t,
+                "priority": "1",
+                "x": str(x),
+                "y": str(y)
+            }
+        )
+
+    return max(len(finputs), len(foutputs))
+
+
+def tile_xml(
+        arch_xml: ET.Element, pbtype_xml: ET.Element, outfile: str,
+        tile_height: int
+):
+    """Generate a top level pb_type containing given pb_type."""
```

Comment:
Need to document return argument, and need to documented that arch_xml is mutated

Reply:
Done.